### DISCUSSION TITLE
Is there file upload with dropzoneJS?

### DISCUSSION THREAD

**Discussion by msharp1** [datenstrom/community#510] [Ask a question] [answered]
Hi,
Somewhere in my project, I plan to add an upload area where user can upload files (not necessarly images). I plan to use [dropzoneJS](https://www.dropzonejs.com/) which sounds pretty solid and I found a [PHP tutorial](https://www.startutorial.com/articles/view/how-to-build-a-file-upload-form-using-dropzonejs-and-php).

I've been able to add dropzone's css and js to load using a custom layout and the drop area to appear. This stage seems to work. Now I need to process the files server side, what would you advise me to put my file `upload.php` ? Do I need to write an extension ? If yes, any pointer to get started is welcome.
Regards, thanks !

**Comment by markseuffert** [comment]
Why not use the built-in file upload? You can configure which users are allowed to do this and which files can be uploaded. The file formats GIF, JPG, MP3, OGG, PDF, PNG, SVG and ZIP are supported per default. https://github.com/annaesvensson/yellow-edit

![screen capture](https://user-images.githubusercontent.com/2429165/39573416-69bdafa4-4ed3-11e8-94db-d1bc1767a4ed.gif)

**Comment by GiovanniSalmeri** [comment]
If for some reason the wise advice regarding the excellent built-in file uploader is not applicable to your case, this is a quick and dirty proof of concept of an extension `dropzone.php`:

```php
<?php
// Dropzone extension
// Copyright (c) 2020 Giovanni Salmeri
// This file may be used and distributed under the terms of the public license.

class YellowDropzone {
    const VERSION = "0.8.10";
    const TYPE = "feature";
    public $yellow;         //access to API
    
    // Handle initialisation
    public function onLoad($yellow) {
        $this->yellow = $yellow;
        $this->yellow->system->setDefault("dropzoneDir", "media/uploads/");
        $this->yellow->system->setDefault("dropzoneAcceptedFiles", ".jpg,.png,.txt,.md,.docx,.rft,.odt,.doc");
        $this->yellow->system->setDefault("dropzoneMaxFilesize", "25"); // in MB
    }
    
    // Handle page content of shortcut
    public function onParseContentShortcut($page, $name, $text, $type) {
        $output = null;
        if ($name=="dropzone" && ($type=="block" || $type=="inline")) {
            $output = "<form action=\"".htmlspecialchars($this->yellow->page->getLocation($absoluteLocation = True))."\" class=\"dropzone\"></form>";
            if (!empty($_FILES)) {
                $tempFile = $_FILES['file']['tmp_name'];
                $targetFile =  $this->yellow->system->get("dropzoneDir").basename($_FILES['file']['name']);
                $fileType = strtoloweru(pathinfo($targetFile, PATHINFO_EXTENSION));
                if ($fileType=="php") $targetFile .= ".disabled";
                $maxFilesize = $this->yellow->system->get("dropzoneMaxFilesize")*1000000;
                $allowedExts = preg_split("/\s*,\s*/", $this->yellow->system->get("dropzoneAcceptedFiles"));
                if (in_array(".".$fileType, $allowedExts) && filesize($tempFile)<=$maxFilesize) move_uploaded_file($tempFile, $targetFile);
                // if the destination file already exists, it will be (silently) overwritten
            }
        }
        return $output;
    }
    
    // Handle page extra data
    public function onParsePageExtra($page, $name) {
        $output = null;
        if ($name == "header") {
            $extensionLocation = $this->yellow->system->get("coreServerBase").$this->yellow->system->get("coreExtensionLocation");
            $maxFilesize = min($this->yellow->toolbox->getNumberBytes(ini_get("post_max_size")), $this->yellow->toolbox->getNumberBytes(ini_get("upload_max_filesize")), $this->yellow->system->get("dropzoneMaxFilesize")*1000000);
            $output = "<link href=\"{$extensionLocation}dropzone.css\" type=\"text/css\" rel=\"stylesheet\" />\n";
            $output .= "<script src=\"{$extensionLocation}dropzone.js\"></script>\n";
            $output .= "<script>Dropzone.prototype.defaultOptions.acceptedFiles = \"".$this->yellow->system->get("dropzoneAcceptedFiles")."\";</script>\n";
            $output .= "<script>Dropzone.prototype.defaultOptions.maxFilesize = \"".$maxFilesize."\";</script>\n";
        }
        return $output;
    }
}
```

I wrote this to show how really easy is to add extensions to Yellow :-) As with most extensions, you have to put a shortcut [dropzone] where you want to appear the drop zone.

Please do not use this draft as is and beware of the **huge** security implications in allowing users to upload arbitrary files! (Read for example https://www.computerweekly.com/answer/File-upload-security-best-practices-Block-a-malicious-file-upload).

Edited 2020-04-20 22:35: fixed two typos that prevented CSS to be read.
Edited 2020-04-21 10:57: code improved, now it is almost usable without huge risks.
Edited 2020-04-22 21:03: $_SERVER['PHP_SELF'] sanitised
Edited 2020-04-23 20:33: see https://github.com/datenstrom/yellow/issues/510#issuecomment-618516155
Edited 2020-04-23 22:30 see https://github.com/datenstrom/yellow/issues/510#issuecomment-618642262

**Comment by BsNoSi** [comment]
> Why not use the built-in file upload? […]

A disadvantage of the existing drag & drop is the fact that everything is copied into the image directory and no structures are possible. I don't know, If "dropzone" offers this – I would not use it because of several other reasons (one of them see security advice preceeding article).

I know that this is no necessity. But I prefer seperating images from zip and/or pdf files. Or to keep all files of a single page in a subfolder which eases maintanance. Especially "grouping" is not that easy with a drag & drop upload. All together in one folder dragging might be easy but what about *later adding* of files?

Creating subfolders for *each* page is no solution, too. Typically, a blog or wiki page has only a single image (in my case) – this would increase the disorder instead of better organization.

Maybe it would be valueble if I can set targets for areas. Dragging files into a page of `content/tools/` moves them to `media/tools` or in a blog copy them into subfolders for year (,month). 

To support a "user upload" I think that there is a absolutely must to seperate uploaded files from the other ones to offer options for checking them before reusing them on the website.

All these options have one general challenge: users need to know what rules apply to the current page. Knowing users I am quite shure that the current technique may be slightly "messy" in pulling everything into `/media/images/` (my very personal opinion) but it is easy to deal with because you don't have to think about path etc. and – you have to have a login to use it. 

@GiovanniSalmeri : You have my great respect and admiration for the quickly created and good looking code.

**Comment by markseuffert** [comment]
Datenstrom Yellow is more powerful that you think. Images and zip/pdf files are uploaded into separate folders, the API allows you to group and check uploaded files in any way necessary.

**Comment by BsNoSi** [comment]
For my personal understanding (means: learning of Yellow techniques):
```
 public function onParsePageExtra($page, $name) {
         $output = null;
         if ($name == "header") {
             $extensionLocation = $this->yellow->system->get("coreServerBase").$this->yellow-system ->get("coreExtensionLocation");
             $output = "<link href=\"{$extensionLocation}dropzone.css\" type=\"text/css\" rel=\"stylesheet\" />\n";
             $output .= "<script defer src=\"{$extensionLocation}dropzone.js\"></script>\n";
```
1. The css and dropzone script is added to *every* page using `header`?
1. Dropzone functionality is available on *every* page using `header`?
2. Pointing to something like `dropzone` would restrict css and javascript to pages using this?
2. What about `bundle.php` – will it combine css into a complete file regardless to that fact that it possibly is needed only on a single page with `dropzone` part?

**Comment by BsNoSi** [comment]
> Datenstrom Yellow is more powerful that you think. […]

To be honest: I thought this. But it needs a lot of php knowledge and time to go through the code to explore the capabilities. Personally, I regret regularly the lack of both…

Can you probably give me a hint where to look to achieve
- uploading files in a subfolder (→ blog files grouped into years)
- seperate "downloadable" files (pdf, zip) into `media/downloads`

**Comment by GiovanniSalmeri** [comment]
I answer for what I understand about the internals of Yellow (one of the beautiful side of contributing to a software project is that one can largely ignore the details of the others' work!)

> 1. The css and dropzone script is added to _every_ page using `header`?
> 2. Dropzone functionality is available on _every_ page using `header`?
> 4. What about `bundle.php` – will it combine css into a complete file regardless to that fact that it possibly is needed only on a single page with `dropzone` part?

Thrice yes. I think that the core developers have chosen this behaviour because normally it *reduces* data traffic: one single bundled file can be cached and anyway is a single http request. Probably there is a way to restrict js and css files to the pages that actually use them, but when the total size is less than a typical small image probably the gain would be anyway negligible.

**Comment by BsNoSi** [comment]
I share this idea for small scripts.  But there are some disadvantages that should be considered:

- If you are using a few *bigger* scripts together with several css for some very special effects each file type combined to one file, this leads to a continuous increase of front page load. 
  - In many cases only a small part of this *preloaded data*  is required there but serveral *welcome images* news ticker a.s.o. lead to a much higher additional load in compare to "usual pages".
  - It slows down response time of the *landing page* 
  - It  increases page load which is relevant for mobile users (typically they have to pay for every bit) therefore should be taken care of.
  - Therefore often a huge amount of data is moved without need for it only to achieve fewer accesses that *might happen* somewhere inside. As a result, a lot of data is sent unnecessarily through the network, which turns the desired speed advantage because of reduced accesses into a slower page load time in general.
  - In high traffic times (Start/End of work,…) smaller files are served much faster than bigger ones leading to a user impression of a »quite fast page« even with a low level server.
- Scripts/functions/style sheets with same name (required for extensions…) but different function (required for different results) collide.
- If you are using extensions that require specific css or js for a desired effect it is harder to achieve this by algorithm instead of a fast written css or js dedicated to the specific page (can be overridden by inline code → increase of traffic).

**To be clear**: *In most cases* **the default is a very good solution**. It is straight forward for creating fast, high quality, results. But if you want to take control over the load of a complex website (that can be easily created and dealed with YELLOW)  it is more complicated to achieve this.

Have I overlooked the option to take control over including or excluding js/css e.g. a header entry `script` or `style` to assign *very special effects* to a single page?

**Comment by GiovanniSalmeri** [comment]
> Have I overlooked the option to take control over including or excluding js/css e.g. a header entry script or style to assign very special effects to a single page?

A javascript solution could be along the lines of <https://stackoverflow.com/questions/950087/how-do-i-include-a-javascript-file-in-another-javascript-file>, with a conditional include. Pseudocode:

```javascript
if (somethingInThisPageRequiresIt()) { import { hugeJavascriptFunction } from "hugeJavascriptModule.js"; }
```

In particular, if I am not mistaken, the method "Dynamic Script Loading" could be used also for CSS. Pseudocode:

```javascript
if (somethingInThisPageRequiresIt()) { 
    var css = document.createElement("link");
    css.href = "urlOfHugeCSS.css";
    css.type = "text/css";
    css.rel = "stylesheet";
    document.head.appendChild(css);
}
```

In this way every page will only include the tiny Javascript code which conditionally loads the actual big script and/or stylesheet.

**Reply by markseuffert** [reply]
Excellent idea. I just found your comment while browsing discussions. If there's enough interest we might be able to handle this in a dynamic bundler pipeline, we could add a configuration that allows dynamic bundling for large CSS/JS resources. Then developers don't have to re-implement it every time. It's on my personal wish list. See also datenstrom/community#289 and #591.

**Comment by msharp1** [comment]
Hi everyone ! my my... ! The thread gone beyond my expectations ;)
Thank you @GiovanniSalmeri for the code you provided and @BsNoSi and @markseu for your insight...

Let me tell a bit more of context of my use case : I'm working on a personal project so as to share audio files. so I need to allow users to upload sound file, eg mp3, wav etc. 

As it is mentioned in some of the comment above, it should not be "open bar", and it needs to be implemented carefully with security in mind.

At a first glance I was thinking not to collect the email of the "contributor" users and keep their privacy intact. What I have in mind is to implement a system like https://mullvad.net/fr/account/#/ do. The user land on a page that provides a random ID, and the user should then upload file(s) with that ID stored somewhere in the data (or filename). If the ID match then the upload is staged. (How we do the staged=>accepted transition is not in the scope of the issue here.)

Of course, I'm pretty sure it's not that secure but ideas are welcome :)

**Comment by msharp1** [comment]
Hi @GiovanniSalmeri ,
I've been trying the extension code. I'm able to make the drop zone to appear with `[dropzone]` tag but then it seems the `POST` action is set to `/yellow/yellow.php` and leads to a **404** error. Am I missing something ?
EDIT - I found that the action path should be `/yellow/dropzone`, that is to say the path to the page named `dropzone` I have created to hold the tag. (my yellow installation root URL is `/yellow/`)

 I'll dig into the API in order to fix the code.

**Comment by GiovanniSalmeri** [comment]
Ops, you are right, I tested the extension only with a local server. With `$_SERVER['REQUEST_URI']` there seems to be the right URL, but there is still something that does not work. I will look into it as soon as possible!

**Comment by msharp1** [comment]
For now I'm using the following code:

```php
$output = "<form action=\"".$this->yellow->page->getLocation($absoluteLocation = True)."\" class=\"dropzone\"></form>";
```

**Comment by GiovanniSalmeri** [comment]
Yes, perfect, thank you @msharp1! Actually it works also with `$_SERVER['REQUEST_URI']` (apparently I was doing earlier something silly...), but your code seems better to me. Remark that the target directory must exist, otherwise the upload will silently fail. I have written some more lines in order to control the behaviour in case of colliding filenames, at this point I think I will publish this little extension in a separate repository.

**Comment by msharp1** [comment]
And I think I spotted a little bug in the following test:

```php
if (in_array(".".$fileType, $allowedExts) && filesize($targetFile)<=$maxFilesize) move_uploaded_file($tempFile, $targetFile);`
```
it seems the above code throw a `PHP Warning:  filesize(): stat failed for media/uploads...` because the file isn't yet moved in the target folder at the time of the test. So I guess the fix is to use `$tempFile` instead for filesize check.

And also `dropzoneLocation` isn't used in the code.

**Comment by GiovanniSalmeri** [comment]
@msharp1 Right, it should be `$tempFile`, thank you!

**Comment by msharp1** [comment]
I am wondering one thing. Let's imagine I want to do more checks on `$tempFile` that are beyond the scope of the _generic_ dropzone plugin (eg. open the file and check specific content...) and this before the file is moved to the final `uploads` folder. Basically, I'd need to call a function but where should I put it ? in another extension ?
Regards,

**Comment by GiovanniSalmeri** [comment]
I'd say that the easiest thing is to include this function in the same extension, for example by putting the additional checks after the line that begins with `$allowedExts = ...`. Anyway I have almost finished a small refactoring of this extension with some little features more, I am going to publish it as soon as possible.

**Comment by GiovanniSalmeri** [comment, marked as answer]
https://github.com/GiovanniSalmeri/yellow-dropzone

Here is the extension. New little features:

+ option for not overwriting files with the same name
+ arguments in shortcut for overriding accepted files and maximum filesize
+ localisation
+ htaccess in uploads directory for security reasons
+ small tweaks in css

This is not yet fully tested. The arguments in shortcut seem to work, but I am not wholly sure that the method I used is robust.
